app/src/main/java/org/mozilla/focus/browser/BrowserFragment.kt

```diff
@@ -268,16 +257,36 @@ class BrowserFragment : EngineViewLifecycleFragment(), Session.Observer {
             }
             else -> {
                 context!!.components.sessionManager.remove()
-                // Delete session, but we allow the parent to handle back behavior.
-                return false
+
+                // We can get into this state in two situations:
+                // - (1) The user is on the "home page" and the overlay is visible. In this situation we just want to
+                //       let the activity handle the back press (and leave the app).
+                // - (2) The overlay is not visible and we are on a website and can't go back further. In this case
+                //       "back" should show the overlay (the next back press closes the app). In this situation we
+                //       cannot look at the URL. We just removed the session and the new session may not be loaded yet.
+                if (!browserOverlay.isVisible) {
+                    // If the browser overlay is not visible then we show it now immediately. Depending on the loading
+                    // state of the page we m
+                    setOverlayVisible(true)
+                } else {
+                    return false
+                }
             }
         }
         return true
     }
 
     fun loadUrl(url: String) {
         if (url.isNotEmpty()) {
-            requireComponents.sessionUseCases.loadUrl.invoke(url)
+            val session = requireComponents.sessionManager.selectedSession
+
+            if (session != null) {
+                // We already have an active session, let's just load the URL.
+                requireComponents.sessionUseCases.loadUrl.invoke(url)
```

Comment:
Nit: Is this logic particular to our use case, or will this be replicated by other projects?  If the latter, could we open an issue to have SessionManager handle it, then add a todo here to replace this when we bump our components version?

Reply:
This is more or less a workaround to get this working with the current logic here.

Ideally I would like this whole back and forth between BrowserFragment, MainActivity and SessionManager.Observer callbacks to disappear: In this app there should always be only one session or none (initial state). Hopefully after the meeting tomorrow I'll know if that's the way to go.